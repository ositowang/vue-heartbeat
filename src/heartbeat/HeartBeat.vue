<template>
  <div :style="cssVariables" class="pulse-wrapper">
    <div
      class="pulse"
      :class="[{'danger':state==='Danger','off': state==='Off'}]"
      :style="{fontSize:fontSize,lineHeight:fontSize}"
    >
      <slot />
      <div v-if="state==='Danger'">
        <slot name="danger">{{dangerText}}</slot>
      </div>
      <div v-else-if="state==='Normal'">
        <slot name="normal">{{normalText}}</slot>
      </div>
      <div v-else>
        <slot name="off">&#x2716;</slot>
      </div>
      <slot/>
    </div>
    <div class="pulse-2" :class="[{off:state==='Off'}]"></div>
  </div>
</template>

<script>
import { convertToUnit } from "../utils";
export default {
  name: "HeartBeat",
  props: {
    width: {
      type: [String, Number],
      default: "20px"
    },
    dangerText: {
      type: [String, Number]
    },
    normalText: {
      type: [String, Number]
    },
    state: {
      type: String,
      default: "Off"
    },
    textSize: {
      type: String,
      default: "10px"
    }
  },
  computed: {
    cssVariables: function() {
      const width =
        typeof this.width === "string" ? parseInt(this.width) : this.width;
      const animationInner = Math.round(width / 2.2);
      const animationOuter = Math.round(width / 1.3);
      let variables = {
        "--circle-width": convertToUnit(this.width),
        "--animation-color": "27, 201, 142",
        "--heartbeat-inner": convertToUnit(animationInner),
        "--heartbeat-outer": convertToUnit(animationOuter)
      };

      if (this.state === "Off") {
        variables["--animation-color"] = "112,120,133";
      } else if (this.state === "Danger") {
        variables["--animation-color"] = "230,71,89";
      }
      return variables;
    },
    fontSize: function() {
      return convertToUnit(this.textSize);
    }
  }
};
</script>

<style lang="scss" scoped>
.pulse-wrapper {
  position: relative;
  width: var(--circle-width);
  height: var(--circle-width);
}
.pulse {
  position: absolute;
  display: flex;
  align-items: center;
  justify-content: center;
  width: var(--circle-width);
  height: var(--circle-width);
  border-radius: 50%;
  background-color: #1bc98e;
  color: #252830;
  animation-name: pulse-1;
  animation-duration: 2s;
  animation-iteration-count: infinite;
  animation-timing-function: cubic-bezier(0.57, 0.14, 0.51, 0.92);
  z-index: 2;
  &.danger {
    background-color: #e64759;
    color: #ffffff;
  }
  &.off {
    background-color: #85898c;
    color: #ffffff;
    animation: none;
  }
}

.pulse-2 {
  position: absolute;
  width: var(--circle-width);
  height: var(--circle-width);
  border-radius: 50%;
  background-color: #1bc98e;
  box-shadow: 0 0 0 rgba(var(--animation-color), 0.25);
  animation-name: pulse-2;
  animation-duration: 2s;
  animation-iteration-count: infinite;
  animation-timing-function: cubic-bezier(0.57, 0.14, 0.51, 0.92);
  z-index: 1;
  &.off {
    animation: none;
  }
}

@keyframes pulse-1 {
  0% {
    -moz-box-shadow: 0 0 0 0 rgba(var(--animation-color), 0.25);
    box-shadow: 0 0 0 0 rgba(var(--animation-color), 0.25);
  }
  70% {
    -moz-box-shadow: 0 0 0 var(--heartbeat-outer)
      rgba(var(--animation-color), 0);
    box-shadow: 0 0 0 var(--heartbeat-outer) rgba(var(--animation-color), 0);
  }
  100% {
    -moz-box-shadow: 0 0 0 0 rgba(var(--animation-color), 0);
    box-shadow: 0 0 0 0 rgba(var(--animation-color), 0);
  }
}

@keyframes pulse-2 {
  0% {
    -moz-box-shadow: 0 0 0 0 rgba(var(--animation-color), 0.25);
    box-shadow: 0 0 0 0 rgba(var(--animation-color), 0.25);
  }
  70% {
    -moz-box-shadow: 0 0 0 var(--heartbeat-inner)
      rgba(var(--animation-color), 0);
    box-shadow: 0 0 0 var(--heartbeat-inner) rgba(var(--animation-color), 0);
  }
  100% {
    -moz-box-shadow: 0 0 0 0 rgba(var(--animation-color), 0);
    box-shadow: 0 0 0 0 rgba(var(--animation-color), 0);
  }
}
</style>