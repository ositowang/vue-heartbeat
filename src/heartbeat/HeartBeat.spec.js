import { shallowMount } from "@vue/test-utils";
import HeartBeat from "./HeartBeat.vue";

let wrapper;

describe("HeartBeat", () => {
  beforeEach(() => {
    wrapper = shallowMount(HeartBeat);
  });

  it("is called", () => {
    expect(wrapper.name()).toBe("HeartBeat");
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it("matches snapshots", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("correctly renders props", () => {
    const text = "Danger";
    wrapper = shallowMount(HeartBeat, {
      propsData: {
        dangerText: text,
        state: "Danger"
      }
    });

    expect(wrapper.text()).toBe(text);
  });

  it("correctly renders state", () => {
    const danger = "Danger";
    wrapper = shallowMount(HeartBeat, {
      propsData: {
        state: danger
      }
    });

    expect(wrapper.findAll(".danger").length).toBe(1);
  });

  it("correctly renders state", () => {
    const off = "Off";
    wrapper = shallowMount(HeartBeat, {
      propsData: {
        state: off
      }
    });

    expect(wrapper.findAll(".off").length).toBe(2);
  });

  it("correctly renders the default slot", () => {
    wrapper = shallowMount(HeartBeat, {
      slots: {
        default: '<div class="default"></div>'
      }
    });

    expect(wrapper.find(".default").html()).toBe('<div class="default"></div>');
  });
});
