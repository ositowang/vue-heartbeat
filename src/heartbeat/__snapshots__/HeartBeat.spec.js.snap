// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeartBeat matches snapshots 1`] = `"<div class=\\"pulse-wrapper\\"><div class=\\"pulse off\\" style=\\"font-size: 20px; line-height: 20px;\\"> <div>âœ–</div> </div> <div class=\\"pulse-2 off\\"></div></div>"`;
